engine/user_preferences.gd

```diff
@@ -0,0 +1,165 @@
+extends Node
+
+# Update this whenever the _user_prefs or _user_controls dictionary are updated
+const schema_version = "0"
+
+# Update this for each release
+const server_version = "0"
+
+const _user_dir = "user://"
+const _save_location = "tetraforce"
+const _preference_file = "preferences.tf"
+
+const default_host = "127.0.0.1"
+
+var _user_controls = {
+	"input.keys": {
+		"UP": KEY_UP,
+		"DOWN": KEY_DOWN,
+		"LEFT": KEY_LEFT,
+		"RIGHT": KEY_RIGHT,
+		"A": KEY_C,
+		"B": KEY_X
+	},
+	"input.axes": {
+		"UP": [JOY_ANALOG_LY, -1.0],
+		"DOWN": [JOY_ANALOG_LY, 1.0],
+		"LEFT": [JOY_ANALOG_LX, -1.0],
+		"RIGHT": [JOY_ANALOG_LX, 1.0]
+	},
+	"input.buttons": {
+		"UP": JOY_DPAD_UP,
+		"DOWN": JOY_DPAD_DOWN,
+		"LEFT": JOY_DPAD_LEFT,
+		"RIGHT": JOY_DPAD_RIGHT,
+		"A": JOY_DS_A,
+		"B": JOY_DS_B
+	}
+}
+
+var _user_prefs = {
+	schema_version = "0",
+	show_name_tags = true,
+	host_address = default_host,
+	display_name = "",
+	skin = 0,
+	controls = _user_controls
+	}
+
+func _ready():
+	var dir = Directory.new()
+	if !dir.dir_exists(_get_save_dir()):
+		dir.open(_user_dir)
+		dir.make_dir(_get_save_dir())
+		
+	get_pref("poo")
+		
+	_load_from_preferences()
+		
+func _get_save_file():
+	return _user_dir + _save_location + _preference_file
+	
+func _get_save_dir():
+	return _user_dir + _save_location
+
+func _load_from_preferences():
+	var saved = File.new()
+		
+	if !saved.file_exists(_get_save_file()):
+		printerr("File not found! Aborting...")
+		_save_to_preferences()
+		
+		return
+
+	saved.open(_get_save_file(), File.READ)
+	
+	var loaded = parse_json(saved.get_as_text())
+	
+	print("loaded save: ", saved.get_as_text())
+	
+	if loaded == null || loaded.empty():
+		_save_to_preferences()
+	elif loaded.schema_version != schema_version:
+		_migrate_save(loaded)
+
+	else:
+		_user_prefs = loaded
+	
+	# uncomment this line if you updated the controls for testing
+#	_user_prefs["controls"] = _user_controls
+	
+	_user_controls = _user_prefs["controls"]
+		
+	saved.close()
+	
+	update_input_map()
+	
+func _save_to_preferences():
+	var saveGame = File.new()
+	saveGame.open(_get_save_file(), File.WRITE)
+
+	var line = JSON.print(_user_prefs)
+	saveGame.store_string(String(line))
+	saveGame.close()
+	
+# Using simple migration for now
+# If we remove a key, currently that data is lost
+# When needed, we can start migrating using the schema numbers to do more complicated migrations
+func _migrate_save(loaded_save):
+	for key in _user_prefs.keys():
+		if loaded_save.has(key):
+			_user_prefs[key] = loaded_save[key]
+	
+	_save_to_preferences()
+	
+func get_pref(key: String):
```

Comment:
I'd rename it to a simple get()

Reply:
get would override behavior in this case.
I agree that get_pref isnt ideal, will think about this and update it

Reply:
eh, good enough then!